File: project_euler/problem_014/solution.py
"""Project Euler Problem 14: Longest Collatz Sequence"""


def solve():

    LIMIT = 1_000_000
    cache = [0] * (LIMIT + 1)
    cache[1] = 1

    for i in range(2, LIMIT):
        n, steps = i, 0
        while n >= LIMIT or cache[n] == 0:
            n = n // 2 if n % 2 == 0 else 3 * n + 1
            steps += 1
        cache[i] = steps + cache[n]

    return max(range(1, LIMIT), key=lambda x: cache[x])


answer = solve()
assert answer == 837799
print(answer)

File: project_euler/problem_014/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    const int N = 1000000;
    vector<int> cache(N + 1, 0);
    cache[1] = 1;

    for (int i = 2; i < N; i++) {
        long long n = i;
        int steps = 0;
        while (n >= N || cache[n] == 0) {
            n = (n % 2 == 0) ? n / 2 : 3 * n + 1;
            steps++;
        }
        cache[i] = steps + (int)cache[n];
    }

    int best = 1;
    for (int i = 2; i < N; i++)
        if (cache[i] > cache[best])
            best = i;

    cout << best << endl;
    return 0;
}
